barcodes_search/wizards/barcode_search.py

```diff
@@ -0,0 +1,121 @@
+        readonly=True)
+
+    @api.multi
+    def button_search(self):
+        self.ensure_one()
+        res = self.search_by_barcode(self.barcode)
+        if len(res) == 1 and not res[0][2]:
+            # Display the result in the according form view
+            record = res[0][1]
+            return record.get_formview_action()
+        else:
+            # Update Wizard Values
+            line_vals = []
+            self.line_ids.unlink()
+            for item in res:
+                line_vals.append((0, 0, {
+                    'field_id': item[0].id,
+                    'item_id': item[1].id,
+                    'item_name': item[1].name,
+                    'extra_data': item[2],
+                }))
+            self.state = 'done'
+            self.line_ids = line_vals
+
+            # Return Action
+            action = self.get_formview_action()
+            action['target'] = 'new'
+            return action
+
+    @api.model
+    def search_by_barcode(self, barcode):
+        """Return a list of tuple (field, record, {extra_data})
+        that matches with the given barcode
+        """
+        barcode_fields = self.get_barcode_fields()
+        res = self._search_by_barcode_barcode_fields(
+            barcode, barcode_fields)
+
+        # Extra search for objects that have a barcode with encoded data
+        # Typical exemple, product with a pattern 20{barcode_base}{NNDDD}
+        # where {NNDDD} is the weight of the product
+        nomenclature_obj = self.env['barcode.nomenclature']
+        nomenclatures = nomenclature_obj.search([])
+        rule_types = self.get_model_by_rule_type()
+        for nomenclature in nomenclatures:
+            parsed_result = nomenclature.parse_barcode(barcode)
+            rule_type = parsed_result['type']
+            if rule_type in rule_types:
+                barcode_fields = self.get_barcode_fields(rule_types[rule_type])
+                base_code = parsed_result['base_code']
+                extra_results = self._search_by_barcode_barcode_fields(
+                    base_code, barcode_fields)
+                for extra_result in extra_results:
+                    extra_result[2]['type'] = rule_type
+                    extra_result[2]['value'] = parsed_result['value']
+                res += extra_results
+        return res
+
+    @api.model
+    def get_barcode_fields(self, model_name=False):
+        """Return a recordset of fields that represent a barcode, in any model.
+        By default, it will return all fields named 'barcode'.
+        Note : Overload that function in a custom module, if you define
+        a barcode field with a name different than 'barcode'"""
+        field_obj = self.env['ir.model.fields']
+        if model_name:
+            domain = [
+                ('name', '=', 'barcode'), ('model', '=', model_name)]
+        else:
+            domain = [
+                ('name', '=', 'barcode'), ('model', '!=', 'barcode.search')]
```

Comment:
I don't think there is ever an instance in which we would actually want `barcode.search` model right? It's transient, plus there is no use case for a barcode on it. In this light, I feel the below would be a bit better:

```python
domain = [
    ('name', '=', 'barcode'), ('model', '!=', 'barcode.search'),
]
if model_name:
    domain.append(('model', '=', model_name))
```

Reply:
Note it might even make sense to add the below at the top of this method, which would help developers if they're being stupid but compile away with a Python running under `PYTHONOPTIMIZE=TRUE`:

```python
assert model_name != 'barcode.search'
```